enrichments/trace/internal/elastic/span.go

```diff
@@ -0,0 +1,296 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package elastic
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	semconv "go.opentelemetry.io/collector/semconv/v1.25.0"
+	"google.golang.org/grpc/codes"
+)
+
+// Span is a representation of OTel `ptrace.Span`. The struct is used
+// to enrich a span with Elastic specific attributes. These attributes
+// are derived from the base attributes and appended to the span
+// attributes. The enrichment logic is performed by categorizing the
+// OTel spans into 2 different categories:
+//   - Elastic transactions, defined as spans which measure the highest
+//     level of work being performed with a service.
+//   - Elastic spans, defined as all spans (including transactions).
+//     However, for the enrichment logic spans are treated as a separate
+//     entity i.e. all transactions are not enriched as spans and vice versa.
+type Span struct {
+	urlFull *url.URL
+
+	peerService     string
+	urlScheme       string
+	urlDomain       string
+	urlPath         string
+	urlQuery        string
+	rpcSystem       string
+	rpcService      string
+	grpcStatus      string
+	dbName          string
+	dbType          string
+	messagingSystem string
+	messagingDestn  string
+
+	urlPort        int64
+	httpStatusCode int64
+
+	spanStatusCode ptrace.StatusCode
+
+	isMessaging        bool
+	isRPC              bool
+	isHTTP             bool
+	isDB               bool
+	messagingDestnTemp bool
+}
+
+func (s *Span) Enrich(span ptrace.Span) {
+	// Extract top level span information.
+	s.spanStatusCode = span.Status().Code()
+
+	// Extract information from span attributes.
+	span.Attributes().Range(func(k string, v pcommon.Value) bool {
+		switch k {
+		case semconv.AttributePeerService:
+			s.peerService = v.Str()
+		case semconv.AttributeMessagingDestinationName:
+			s.isMessaging = true
+			s.messagingDestn = v.Str()
+		case semconv.AttributeMessagingOperation:
+			s.isMessaging = true
+		case semconv.AttributeMessagingSystem:
+			s.isMessaging = true
+			s.messagingSystem = v.Str()
+		case semconv.AttributeMessagingDestinationTemporary:
+			s.isMessaging = true
+			s.messagingDestnTemp = true
+		case semconv.AttributeHTTPStatusCode,
+			semconv.AttributeHTTPResponseStatusCode:
+			s.isHTTP = true
+			s.httpStatusCode = v.Int()
+		case semconv.AttributeHTTPMethod,
+			semconv.AttributeHTTPRequestMethod,
+			semconv.AttributeHTTPURL,
+			semconv.AttributeHTTPTarget,
+			semconv.AttributeHTTPScheme,
+			semconv.AttributeHTTPFlavor,
+			semconv.AttributeNetHostName:
+			s.isHTTP = true
+		// TODO (lahsivjar): Is url ONLY present for HTTP? This is what is done in apm-data.
```

Comment:
Ahh, that's the next huge topic we need to think about :) That is versioning across different SemConv versions.

`db.name` got renamed to `db.namespace` - can be [seen here](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/model/registry/deprecated/db.yaml#L69-L74). I see Elasticsearch in v1.26.0 which you linked doesn't have `db.namespace`, but that seems to be an error - [`main` already has it](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/elasticsearch.md) - that's how it should be and `db.namespace` be there in the next version.

This happens as part of the DB stabilization - currently all db attributes are experimental, the next version aims to make most of these stable, but as part of this, there are some changes.

As it seems in SemConv v1.26.0 the renaming of `db.name` to `db.namespace` already happened (except for ES, where it just got removed) and got released in the SemConv spec, but that's just the SemConv spec, I doubt any real implementation implemented that. I'd expect most still send `db.name` - in case of Elasticsearch I'm fairly confident we haven't updated this yet (I in discussion wit the ES clients team to go through the SemConv changes and implement the new version - I'm not aware they implemented any of this yet).

> OTOH, vendor-specific attributes like, [db.elasticsearch.cluster.name](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/docs/attributes-registry/db.md), feel a bit brittle to handle.

Agree, that can explode quickly and the goal of SemConv is to have common attributes across vendors (and only have vendor specific ones where it's really needed), so we should take advantage of this and use the vendor agnostic attribute.

> Currently, the logic to determine service target uses db.name but if such attributes are not expected in the semconv maybe we should expand the definition.

Ultimately I think we should have a better story for version handling. In this case, depending on the version of the schema url that's sent, it should be either `db.name` (old version) or `db.namespace` from version `1.26` and higher.

To unblock the PR: since I think no instrumentation implemented `db.namespace` yet, I think having `db.name` for now is fine. And then I'd handle versioning in an extra issue/PR later. 

Alternatively you could check for both - that'd work and one should be always present.

Reply:
> db.name got renamed to db.namespace - can be [seen here](https://github.com/open-telemetry/semantic-conventions/blob/v1.26.0/model/registry/deprecated/db.yaml#L69-L74). I see Elasticsearch in v1.26.0 which you linked doesn't have db.namespace, but that seems to be an error - [main already has it](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/elasticsearch.md) - that's how it should be and db.namespace be there in the next version.

Hmm, I see. I have been working with v1.25.0 of the semconv (the latest released one with v0.104.0) so I didn't even see that db.name has been deprecated. Thanks for pointing me to it.

> To unblock the PR: since I think no instrumentation implemented db.namespace yet, I think having db.name for now is fine. And then I'd handle versioning in an extra issue/PR later.

> Alternatively you could check for both - that'd work and one should be always present.

It would probably be best to handle this in a followup PR after we have the v1.26.0 of the semconv released.